File: src/2025-02-24-deepclaude-rust/cpp/exercises/practice/phone-number/phone_number.cpp
#include "phone_number.h"

namespace phone_number {

phone_number::phone_number(const std::string& input) {
    std::string cleaned = std::regex_replace(input, std::regex(R"([^\d])"), "");
    
    if (cleaned.length() == 11) {
        if (cleaned[0] != '1') {
            throw std::invalid_argument("Country code must be 1");
        }
        cleaned = cleaned.substr(1);
    }
    
    if (cleaned.length() != 10) {
        throw std::invalid_argument("Invalid number of digits");
    }
    
    if (cleaned[0] < '2' || cleaned[0] > '9') {
        throw std::invalid_argument("Area code cannot start with 0 or 1");
    }
    
    if (cleaned[3] < '2' || cleaned[3] > '9') {
        throw std::invalid_argument("Exchange code cannot start with 0 or 1");
    }
    
    number_ = cleaned;
}

std::string phone_number::number() const {
    return number_;
}

std::string phone_number::area_code() const {
    return number_.substr(0, 3);
}

}  // namespace phone_number

File: src/2025-02-24-deepclaude-rust/python/exercises/practice/phone-number/phone_number.py
class PhoneNumber:
    def __init__(self, number):
        # Check for letters
        if any(c.isalpha() for c in number):
            raise ValueError("letters not permitted")
        
        # Check for invalid punctuation
        allowed_punctuation = {'+', '-', '.', ' ', '(', ')'}
        for c in number:
            if not c.isdigit() and c not in allowed_punctuation:
                raise ValueError("punctuations not permitted")
        
        # Extract digits
        digits = ''.join(filter(str.isdigit, number))
        
        # Check length
        if len(digits) < 10:
            raise ValueError("must not be fewer than 10 digits")
        if len(digits) > 11:
            raise ValueError("must not be greater than 11 digits")
        
        # Handle country code
        if len(digits) == 11:
            if digits[0] != '1':
                raise ValueError("11 digits must start with 1")
            digits = digits[1:]
        
        # Check area code
        area_code = digits[0:3]
        if area_code[0] == '0':
            raise ValueError("area code cannot start with zero")
        if area_code[0] == '1':
            raise ValueError("area code cannot start with one")
        
        # Check exchange code
        exchange_code = digits[3:6]
        if exchange_code[0] == '0':
            raise ValueError("exchange code cannot start with zero")
        if exchange_code[0] == '1':
            raise ValueError("exchange code cannot start with one")
        
        self.number = digits
        self.area_code = digits[0:3]
        self.exchange_code = digits[3:6]
        self.subscriber_number = digits[6:10]

    def pretty(self):
        return f"({self.area_code})-{self.exchange_code}-{self.subscriber_number}"
